// @applitools/driver/dist/capabilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseCapabilities = void 0;
function parseCapabilities(capabilities) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    if (capabilities.capabilities)
        capabilities = capabilities.capabilities;
    const info = {
        browserName: !capabilities.app && !capabilities.bundleId
            ? ((_a = capabilities.browserName) !== null && _a !== void 0 ? _a : (_b = capabilities.desired) === null || _b === void 0 ? void 0 : _b.browserName) || undefined
            : undefined,
        browserVersion: ((_c = capabilities.browserVersion) !== null && _c !== void 0 ? _c : capabilities.version) || undefined,
        platformName: ((_e = (_d = capabilities.platformName) !== null && _d !== void 0 ? _d : capabilities.platform) !== null && _e !== void 0 ? _e : (_f = capabilities.desired) === null || _f === void 0 ? void 0 : _f.platformName) || undefined,
        platformVersion: capabilities.platformVersion || undefined,
        isW3C: isW3C(capabilities),
        isMobile: isMobile(capabilities),
    };
    if (info.isMobile) {
        info.deviceName = ((_h = (_g = capabilities.desired) === null || _g === void 0 ? void 0 : _g.deviceName) !== null && _h !== void 0 ? _h : capabilities.deviceName) || undefined;
        info.isNative = info.isMobile && !info.browserName;
        info.isIOS = isIOS(capabilities);
        info.isAndroid = isAndroid(capabilities);
        info.orientation = (_k = ((_j = capabilities.deviceOrientation) !== null && _j !== void 0 ? _j : capabilities.orientation)) === null || _k === void 0 ? void 0 : _k.toLowerCase();
    }
    if (info.isNative) {
        info.pixelRatio = capabilities.pixelRatio;
        info.statusBarHeight = capabilities.statBarHeight;
    }
    return info;
}
exports.parseCapabilities = parseCapabilities;
function isW3C(capabilities) {
    const isW3C = Boolean((capabilities.platformName || capabilities.browserVersion) &&
        (capabilities.platformVersion || capabilities.hasOwnProperty('setWindowRect')));
    return isW3C || isAppium(capabilities);
}
function isAppium(capabilities) {
    return (Boolean(capabilities.automationName || capabilities.deviceName || capabilities.appiumVersion) ||
        Object.keys(capabilities).some(cap => cap.startsWith('appium:')));
}
function _isChrome(capabilities) {
    return Boolean(capabilities.chrome || capabilities['goog:chromeOptions']);
}
function _isFirefox(capabilities) {
    return capabilities.browserName === 'firefox' || Object.keys(capabilities).some(cap => cap.startsWith('moz:'));
}
function isMobile(capabilities) {
    var _a, _b;
    return (capabilities.browserName === '' ||
        ['ipad', 'iphone', 'android'].includes((_b = (_a = capabilities.browserName) === null || _a === void 0 ? void 0 : _a.toLowerCase()) !== null && _b !== void 0 ? _b : '') ||
        isAppium(capabilities));
}
function isIOS(capabilities) {
    return /iOS/i.test(capabilities.platformName) || /(iPad|iPhone)/i.test(capabilities.deviceName);
}
function isAndroid(capabilities) {
    return /Android/i.test(capabilities.platformName) || /Android/i.test(capabilities.browserName);
}
